File: Queues/QueuesInJava/queues.py
class ListItem:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next


class Queue:
    def __init__(self):
        self.front = None
        self.back = None

    def enqueue(self, item):
        value = ListItem(item)

        if (self.front == None):
            self.front = self.back = value
        else:
            self.back.next = value
            self.back = value

    def dequeue(self):
        if (self.front == None):
            raise Exception('The Queue is empty')

        if(self.front == self.back):
            self.front = self.back = None
            return
        else:
            second = self.front.next
            self.front = None
            self.front = second

        return self.front.value

    def peek(self):
        return self.front.value

    def isEmpty(self):
        return (self.front == None)

File: Queues/ExerciseReversingAQueue/queues.js
class ListItem{
    constructor(value=null, next=null){
        this.value = value;
        this.next = next;
    }
}

class Queue{
    constructor(front=null, back=null){
        this.front = front;
        this.back = back;
    }

    enqueue(item){
        let value = new ListItem(item);

        if (this.front == null){
            this.front=this.back= value;
        } else{
            this.back.next = value;
            this.back = value
        }
    }

    dequeue(){
        if (this.front == null) throw new Error('The Queue is empty');

        let previous = this.front;
        
        if(this.front == this.back){
            this.front=this.back=null;
        }else{
            let second = this.front.next;
            this.front = null;
            this.front = second;
        }

        return previous.value;
    }

    peek(){
        if (this.front == null) throw new Error('The Queue is empty');
        return this.front.value;
    }

    isEmpty(){
        return (this.front == null);
    }

    getQueue(){
        let array = [];
        while (this.front != null){
            array.push(this.dequeue())

        }

        return array;
    }
}

module.exports = Queue;